tests/components/zha/test_discover.py

```diff
@@ -30,45 +66,298 @@
+
+    ep_channels = mock.MagicMock(spec_set=zha_channels.ChannelPool)
+    ep_mock = mock.PropertyMock()
+    ep_mock.return_value.profile_id = 0x0104
+    ep_mock.return_value.device_type = 0x0100
+    type(ep_channels).endpoint = ep_mock
+
+    overrides = {ep_channels.unique_id: {"type": zha_const.SWITCH}}
+    get_entity_mock = mock.MagicMock(
+        return_value=(mock.sentinel.entity_cls, mock.sentinel.claimed)
+    )
+    with mock.patch(
+        "homeassistant.components.zha.core.registries.ZHA_ENTITIES.get_entity",
+        get_entity_mock,
+    ):
+        with mock.patch.dict(disc.PROBE._device_configs, overrides, clear=True):
+            disc.PROBE.discover_by_device_type(ep_channels)
+            assert get_entity_mock.call_count == 1
+            assert ep_channels.claim_channels.call_count == 1
+            assert ep_channels.claim_channels.call_args[0][0] is mock.sentinel.claimed
+            assert ep_channels.async_new_entity.call_count == 1
+            assert ep_channels.async_new_entity.call_args[0][0] == zha_const.SWITCH
+            assert (
+                ep_channels.async_new_entity.call_args[0][1] == mock.sentinel.entity_cls
+            )
+
+
+def test_discover_probe_single_cluster():
+    """Test entity discovery by single cluster."""
+
+    ep_channels = mock.MagicMock(spec_set=zha_channels.ChannelPool)
+    ep_mock = mock.PropertyMock()
+    ep_mock.return_value.profile_id = 0x0104
+    ep_mock.return_value.device_type = 0x0100
+    type(ep_channels).endpoint = ep_mock
+
+    get_entity_mock = mock.MagicMock(
+        return_value=(mock.sentinel.entity_cls, mock.sentinel.claimed)
+    )
+    channel_mock = mock.MagicMock(spec_set=base_channels.ZigbeeChannel)
+    with mock.patch(
+        "homeassistant.components.zha.core.registries.ZHA_ENTITIES.get_entity",
+        get_entity_mock,
+    ):
+        disc.PROBE.probe_single_cluster(zha_const.SWITCH, channel_mock, ep_channels)
+
+    assert get_entity_mock.call_count == 1
+    assert ep_channels.claim_channels.call_count == 1
+    assert ep_channels.claim_channels.call_args[0][0] is mock.sentinel.claimed
+    assert ep_channels.async_new_entity.call_count == 1
+    assert ep_channels.async_new_entity.call_args[0][0] == zha_const.SWITCH
+    assert ep_channels.async_new_entity.call_args[0][1] == mock.sentinel.entity_cls
+    assert ep_channels.async_new_entity.call_args[0][3] == mock.sentinel.claimed
+
+
+@pytest.mark.parametrize("device_info", DEVICES)
+async def test_discover_endpoint(device_info, channels_mock, hass):
+    """Test device discovery."""
+
+    with mock.patch(
+        "homeassistant.components.zha.core.channels.Channels.async_new_entity"
+    ) as new_ent:
+        channels = channels_mock(
+            device_info["endpoints"],
+            manufacturer=device_info["manufacturer"],
+            model=device_info["model"],
+            node_desc=device_info["node_descriptor"],
+        )
+
+    assert device_info["event_channels"] == sorted(
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5051600)